7.0 account invoice unique customer ref

mcassuto opened the pull request:
This PR add 2 modules: 
1/ unique_supplier_invoice_number which set a constraint on (partner_id, supplier_invoice_number) to be unique
2/ account_invoice_unique_customer_ref which set a constraint on (partner_id, name) to be unique, only for out_invoice or out_refund type account_invoice.

Commit:
Add module unique_supplier_invoice_number

Commit:
unique_supplier_invoice_number: Fix missing import of account_invoice.

Commit:
unique_supplier_invoice_number: fix duplicate -remove field supplier_invoice_number.

Commit:
unique_supplier_invoice_number : update description + fix unicity per partner

Commit:
account_invoice_unique_customer_ref : New module !

rvalyi commented on account_invoice_unique_customer_ref/account_invoice.py:
```diff
@@ -0,0 +1,67 @@
+        if invoice_type not in ['out_invoice', 'out_refund']:
+            return True
+
+        sr_ids = self.search(cr, uid, [("type", "=", invoice_type), ("partner_id", "=", invoice_partner.id)],
```
line too long; max 80

rvalyi commented on account_invoice_unique_customer_ref/account_invoice.py:
```diff
@@ -0,0 +1,67 @@
+#
+# OpenERP, Open Source Management Solution
+# This module copyright (C) 2010 - 2014 Savoir-faire Linux
+# (<http://www.savoirfairelinux.com>).
```
missing 4 spaces at the beginning of these 3 lines

rvalyi commented on account_invoice_unique_customer_ref/account_invoice.py:
```diff
@@ -0,0 +1,67 @@
+        return super(account_invoice, self).copy(cr, uid, id, default, context)
+
+    def _check_unique_name_insensitive(self, cr, uid, ids, context=None):
+        # this function only works with one ids
```
"one ids" ? -> "one id"  or "an ids array of a single id"

rvalyi commented on unique_supplier_invoice_number/account_invoice.py:
```diff
@@ -0,0 +1,63 @@
+
+    def _check_unique_supplier_invoice_number_insensitive(self, cr, uid, ids,
+                                                          context=None):
+        # this function only works with one ids
```
"one ids" -> same remark

rvalyi commented:
Hello I made some non exhaustive comments.

Also, are we accepting a company logo in OCA modules? I'm all in favor that modules authors/companies get the visibility that fuel the work, but I think till now we were not accepting company logos. If we do, we should see how we do it (I'm not against but only if it is highly moderated).

elicoidal commented:
No company logo or advertisement accepted.
Of course the authoring can/has to mention companies

Eric CAUDAL

Eric Caudal

On 07/26/2014 12:29 PM, Raphaël Valyi wrote:

> Hello I made some non exhaustive comments.
> 
> Also, are we accepting a company logo in OCA modules? I'm all in favor 
> that modules authors/companies get the visibility that fuel the work, 
> but I think till now we were not accepting company logos. If we do, we 
> should see how we do it (I'm not against but only if is highly moderated).
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/OCA/account-invoicing/pull/2#issuecomment-50223028.

pedrobaeza commented:
Hi, one remark without making exhaustive review: first module name must follow conventions about names, as you have made on the seconds, so I propose _account_invoice_supplier_ref_unique_ or _account_invoice_supplier_number_unique_.

Regards.

bwrsandman commented:
@mcassuto, the sfl logo is a placeholder.
You're supposed to look for some creative commons logos which match the module itself.

Commit:
Fix after all the reviews :
• rename module unique_supplier_invoice_number to account_invoice_unique_supplier_invoice_number
• remove logos
• add .pot file to both modules  account_invoice_unique_supplier_invoice_number and account_invoice_unique_customer_ref
• add fr translation to both modules
• fix typo '# this function only works with one id' in both modules
• fix the missing spaces in the licence text for both modules
• fix the too long lines

mcassuto commented:
All fixed:
• rename module unique_supplier_invoice_number to account_invoice_unique_supplier_invoice_number
• remove logos
• add .pot file to both modules  account_invoice_unique_supplier_invoice_number and account_invoice_unique_customer_ref
• add fr translation to both modules
• fix typo '# this function only works with one id' in both modules
• fix the missing spaces in the licence text for both modules
• fix the too long lines"

rvalyi commented on account_invoice_unique_supplier_invoice_number/account_invoice.py:
```diff
@@ -0,0 +1,66 @@
+            self.browse(cr, uid, sr_ids, context=context)
+            if x.supplier_invoice_number and x.id not in ids
+        ]
+        if invoice.supplier_invoice_number and invoice.supplier_invoice_number.lower() in lst:
```
line is 94 chars, limit will soon be 80 chars in Travis tests
You can beat @bwrsandman if you disagree but this is were OCA stands so far.

rvalyi commented on account_invoice_unique_customer_ref/account_invoice.py:
```diff
@@ -0,0 +1,69 @@
+# #############################################################################
+#
+# OpenERP, Open Source Management Solution
+# This module copyright (C) 2010 - 2014 Savoir-faire Linux
```
bad indent on the 2 upper lines

pedrobaeza commented:
Sorry for being too picky, but the name you have chosen can be problematic due to its length: there is a restriction on the size for XML-ID names, and you already catch 47 with this name, what leaves only 17 chars for the rest of the ID. That's why I proposed you the other names. There is also a redundancy in your name, using invoice two times. Following the other module name, why don't you name it _account_invoice_unique_supplier_ref_?

Regards.

Commit:
All files Pep8 compliant

Commit:
Rename module name to respect  XML-ID names restriction

mcassuto commented:
@pedrobaeza : modules renamed
@rvalyi : all files checked and PEP8 complient

pedrobaeza commented:
One last thing: you have to remove .pyc and .mo files.

Regards.

Commit:
Clean .pyc & .mo files

mcassuto commented:
@pedrobaeza : done ;-)

pedrobaeza commented:
Thanks!

:+1:

mcassuto commented:
@pedrobaeza : just for the understanding, what about having these modules merged ?

pedrobaeza commented:
We need at least one approval more from another reviewer, so please wait for it. If we see that nobody can review for a while, we can measure its inclussion.

Regards.

mcassuto commented:
thanks a lot Pedro @pedrobaeza

rvalyi commented:
@mcassuto 

sorry if that's a dumb question,  but your _check_unique_\* methods work with only the 1st id of the ids collection. What happens if you write several invoices at once, with some API call or via some RPC?

Wouldn't it be better to loop over all ids and enforce the test on all ids? Can you justify why you take only the 1st element?

bwrsandman commented on account_invoice_customer_ref_unique/account_invoice.py:
```diff
@@ -0,0 +1,71 @@
+
+    def _check_unique_name_insensitive(self, cr, uid, ids, context=None):
+        # this function only works with one id
+        if ids:
```
`if ids and len(ids) == 1:`

bwrsandman commented on account_invoice_supplier_ref_unique/account_invoice.py:
```diff
@@ -0,0 +1,67 @@
+    def _check_unique_supplier_invoice_number_insensitive(self, cr, uid, ids,
+                                                          context=None):
+        # this function only works with one id
+        if ids:
```
`if ids and len(ids) == 1:`

bwrsandman commented on account_invoice_supplier_ref_unique/account_invoice.py:
```diff
@@ -0,0 +1,67 @@
+            return False
+        return True
+
+    msg = _(
```
do here what you did [here](https://github.com/OCA/account-invoicing/pull/2/files#diff-ea02e43001b9562a6ca6660e63620aaeR66)

mcassuto commented:
@rvalyi : because the id is provided by postgres.... hence it is almost impossible to fall into this case.
We had this discussion with another developer internally and we added this check just to enforce this code robustness.

Commit:
account_invoice_supplier_ref_unique : make constraint translation to work with different langage within same instance + add bwrsandman code suggestion

mcassuto commented:
@bwrsandman : my bad for the langage translation... I forgot. 
Other comment added

rvalyi commented:
:+1:  ok thanks for justifying. Accepted.

max3903 commented:
:+1:

bwrsandman commented on account_invoice_supplier_ref_unique/account_invoice.py:
```diff
@@ -0,0 +1,71 @@
+        return True
+
+    def _rec_message(self, cr, uid, ids, context=None):
+        return _('TThe supplier invoice number must be unique \
```
> TThe

Commit:
account_invoice_supplier_ref_unique : Fix msg typo

mcassuto commented:
@bwrsandman : typo fixed and checked in pot/po files

bwrsandman commented on account_invoice_supplier_ref_unique/account_invoice.py:
```diff
@@ -0,0 +1,71 @@
+        else:
+            raise orm.except_orm(
+                _('Error'),
+                'Cannot check unique supplier ref without id.')
```
This string isn't translated

bwrsandman commented on account_invoice_supplier_ref_unique/account_invoice.py:
```diff
@@ -0,0 +1,71 @@
+    _constraints = [
+        (_check_unique_supplier_invoice_number_insensitive,
+         _rec_message,
+         ['name'])
```
Unless you add `'supplier_invoice_number'` to the list, the constraint will only be checked when name changes...

bwrsandman commented on account_invoice_customer_ref_unique/account_invoice.py:
```diff
@@ -0,0 +1,71 @@
+            i_id = ids[0]
+        else:
+            raise orm.except_orm(_('Error'),
+                                 'Cannot check unique name without id.')
```
This string isn't translated

bwrsandman commented:
:-1: 
Branch is in PR against the wrong branch (master)
You will need to close this one and make a new PR with the same branch.
Here's a link to do it quickly: https://github.com/savoirfairelinux/account-invoicing/compare/OCA:7.0...savoirfairelinux:7.0_account_invoice_unique_customer_ref

bwrsandman commented:
@max3903 @rvalyi @pedrobaeza, please check the target branches before approving.

mcassuto commented:
Pull request made on the wrong branch

rvalyi commented:
sorry

pedrobaeza commented:
Sorry for the mistake, but these modules can also apply to 8.0, doesn't it?
